var/spack/repos/builtin/packages/py-pytorch/package.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+
+class PyPytorch(PythonPackage):
+    """Tensors and Dynamic neural networks in Python
+    with strong GPU acceleration."""
+
+    homepage = "http://pytorch.org/"
+    url      = "https://github.com/pytorch/pytorch/archive/v0.3.1.tar.gz"
+
+    version('0.4.0', git='https://github.com/pytorch/pytorch.git',
+            tag='v0.4.0', submodules=True)
+    version('0.3.1', git='https://github.com/pytorch/pytorch.git',
+            tag='v0.3.1', submodules=True)
+
+    variant('cuda', default='False', description='Add GPU support')
+    variant('cudnn', default='False', description='Add cuDNN support')
+    variant('nccl', default='False', description='Add NCCL support')
+    variant('mkldnn', default='False', description='Add Intel MKL DNN support')
+    variant('magma', default='False', description='Add MAGMA support')
+
+    conflicts('+cudnn', when='~cuda')
+    conflicts('+nccl', when='~cuda')
+    conflicts('+magma', when='~cuda')
+    conflicts('+mkldnn', when='@:0.3.2')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-cffi', type='build')
+    depends_on('py-numpy', type=('run', 'build'))
+    depends_on('blas', type='build')
+    depends_on('lapack', type='build')
+    depends_on('py-pyyaml', type=('run', 'build'))
+    depends_on('py-typing', when='@0.3.2:', type=('run', 'build'))
+    depends_on('intel-mkl', when='+mkl', type='build')
+    depends_on('cuda', when='+cuda', type=('build', 'run'))
+    depends_on('cudnn', when='+cuda+cudnn', type='build')
+    depends_on('nccl', when='+cuda+nccl', type='build')
+    depends_on('magma+shared', when='+cuda+magma', type='build')
+
+    def setup_environment(self, build_env, run_env):
+        build_env.set('MAX_JOBS', make_jobs)
+
+        if '+cuda' in self.spec:
```

Comment:
these should be `self.spec.satisfies('+cuda')`, which only looks at the variants of this package, your version technically searches the complete DAG.

Reply:
We should either make this syntax work the way people expect or remove it from the documentation and the 99% of Spack packages that use it. @alalazo

Reply:
See #4135. If we can decide what is the semantic (`could_satisfy` vs. `satisfy` vs. `__contains__`) I can try to submit a PR. 

Also: I think that this kind of fix must be coordinated carefully with @tgamblin or any other owner of Spack as:
1. The PR will touch *a lot of files* (and that's an understatement).
2. Maintaining it open for a long time and rebase or update it consistently will be a major pain.

@healther It shouldn't be:
```python
self.spec.satisfies('+cuda')
```
but
```python
self.spec.satisfies('+cuda', strict=True)
```
unless the behavior changed since #4135.

Reply:
Oh, thanks for the correction there, I don't think that I ever used it for non-bool variants. Then we should probably finish the discussion of #4135.

Reply:
As far as this PR goes, since every other package in Spack uses the `if '+foo' in spec` syntax, I think this is fine for now. We'll fix everything in another PR.

Reply:
I don't mind doing the changes that would one less to do later. I just have to know what is the proper way, with `strict=True` or it is not necessary ?

Reply:
I'm not sure what the proper way to do things will be. It doesn't seem like a consensus was reached in #4135, but @tgamblin seems to be in favor of making `if '+foo' in spec` work the way one would expect, so your current syntax is most likely to last.